### PR TITLE
reduce parameter size for finite difference operators

### LINKED ISSUE #1257: Apply GPU parameter space fix to finite difference operators

### PR DESCRIPTION
**Pull request by simonbyrne**
Fixes #1257 

- [x] Code follows the [style guidelines](https://clima.github.io/ClimateMachine.jl/latest/DevDocs/CodeStyle/) OR N/A.
- [x] Unit tests are included OR N/A.
- [x] Code is exercised in an integration test OR N/A.
- [x] Documentation has been added/updated OR N/A.

### CODE REVIEW

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by charleskawczynski** [comment]
Maybe this would be better for a separate PR, but wouldn't all the return type pieces be easier if we packed the operator type into the type hierarchy so that we can write general rules for the return spaces:

```julia
# module Spaces

dual_space(space::CenterFiniteDifferenceSpace) = FaceFiniteDifferenceSpace(space)
dual_space(space::FaceFiniteDifferenceSpace) = CenterFiniteDifferenceSpace(space)
dual_space(space::CenterExtrudedFiniteDifferenceSpace) = FaceExtrudedFiniteDifferenceSpace(space)
dual_space(space::FaceExtrudedFiniteDifferenceSpace) = CenterExtrudedFiniteDifferenceSpace(space)

# module Operators
abstract type FiniteDifferenceOperator{OpType} <: AbstractOperator end

struct Staggered end
struct Collocated end

return_space(::FiniteDifferenceOperator{Staggered}, space) = Spaces.dual_space(space)
return_space(::FiniteDifferenceOperator{Collocated}, space) = Spaces.dual_space(space)
```

And then use, for example, `struct InterpolateF2C{BCS} <: InterpolationOperator{Staggered}`.

The annoying part of that I guess is the weighted and advection ops, so we would need to handle a few more cases.

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

**Comment by simonbyrne** [comment]
bors try

File: src/DataLayouts/DataLayouts.jl [hunk 1]
```diff
@@ -491,6 +491,10 @@ end
 @inline function Base.getindex(data::IFH{S}, i, _, _, _, h) where {S}
     @inbounds get_struct(parent(data), S, Val(2), CartesianIndex(i, 1, h))
 end
+@inline function Base.getindex(data::IFH{S}, I::CartesianIndex{5}) where {S}
```
Comment (charleskawczynski) [inline src/DataLayouts/DataLayouts.jl:494]: I'm really happy about these additions (which, IIUC, were added due to removing `getidx(column(field, hidx...), loc, idx)` below). I think this is generally a better memory access pattern, it reminds me of performance benefits of `a[i,j]` over `a[i][j]`.

**Review by charleskawczynski — APPROVED** [review]

**Comment by charleskawczynski** [comment]
bors ping

**Pull request merged**